### DISCUSSION TITLE
Dynamic parameters

### DISCUSSION THREAD

**Discussion by ypiel-talend** [verhas/jamal#23] [Q&A] [answered]
Hi,

In documentation, it seems that all examples have static list of paramters. For instance with @for: https://github.com/verhas/jamal#for

> {@for variable in (a,b,c,d)= content to be repeated
> containing variable}

Is it possible to not have fixed values (a, b, c, d) but, read those values from another file ? Each line would be given as parameter ?
Example:

> {@for $a in (read myFile)= Line is $a
> }

**Comment by verhas** [comment, marked as answer]
I like that you wrote:

"it seems that all examples have static list..."

I do because they are not. They may seem, but in Jamal, nothing is static. Anything can be the result of a macro evaluation. So if you have a file and want to have a `for` to iterate through all the lines, you need two things, plus an optional third:

1. Specify that the value separator is a new line

2. Somehow, get the lines between the opening `(` and closing `)` strings.

3. Use something else other than `( ` and `)` as opening and closing strings when the `)` character may be present in the file.

How should you do that? Let's get it one by one.

The `for` macro has an option called `separator`. For example, the following code:
```
{@for [separator="\n"] $i in (bear
zebra
snake
dog
cat)=* $i
}
```
It will use the words as elements and generate the itemized list of the animals. Note that empty lines are just like regular lines. If you want to skip the blank lines, you can also use the `skipEmpty` option.

To get the content of a file, you should use the macro `include`. I copied the names of the animals into an external file named `animals.txt`, and then I could use the following code:

```
{#for [separator="\n" skipEmpty] $i in ({@include [verbatim] animals.txt})=* $i
}
```

Note that using `{@for...` would not work because the content of the `for` macro, specifically the include, has to be evaluated before the for itself.

It is often only the list of values that is the part to be evaluated, but the content repeated many times must not. Therefore, the second part must be protected from an evaluation using the `ident` macro or the `escape` macro. In this example, the `* $i` is evaluated, and it is not a problem. It just evaluates to itself. In other cases, it should be written as

```
{#for [separator="\n" skipEmpty] $i in ({@include [verbatim] animals.txt})={@ident * $i}
}
```

The situation that the value list IS to be evaluated but the repeated text is not is so frequent that the `for` macro supports this with a unique option `evaluateValueList` or `evalist` for short. If this option is used, the `for` macro itself will evaluate the value list before starting the iterations. That way, it could be

```
{@for [separator="\n" skipEmpty evalist] $i in ({@include [verbatim] animals.txt})=* $i
}
```
But at the moment, it does not work due to a bug that I will fix in the next release. (In git you can see the commit with the fix.)

The last thing is that you want to use something else instead of `(` and `)`. Without repeating what is written in the documentation, your example may look like the following:

```
{#for [separator="\n" skipEmpty] $i in `--FILE INCLUDE BOUNDARY`
{@include [verbatim] animals.txt}
`--FILE INCLUDE BOUNDARY`=* $i
}
```

**Reply by ypiel-talend** [reply]
Thanks a lot for your answer.

But I can't exec your example:

> ~/dev/github/perso/pocjamal$ cat for.out.jam 
> {@for [separator="\n"] $i in (bear
> zebra
> snake
> dog
> cat)=* $i
> }

> ~/dev/github/perso/pocjamal$ jshell https://raw.githubusercontent.com/verhas/jamal/master/jamal-cmd/jamal.jshell

> [INFO] URL=https://repo.maven.apache.org/maven2/com/javax0/jamal/jamal-engine/1.6.5/jamal-engine-1.6.5.jar                                                                                                 
> [INFO] /home/ypiel/.m2/repository/com/javax0/jamal/jamal-engine/1.6.5/jamal-engine-1.6.5.jar is already in the repo                                                                                        
> [INFO] URL=https://repo.maven.apache.org/maven2/com/javax0/jamal/jamal-api/1.6.5/jamal-api-1.6.5.jar                                                                                                       
> [INFO] /home/ypiel/.m2/repository/com/javax0/jamal/jamal-api/1.6.5/jamal-api-1.6.5.jar is already in the repo 
> ..........
> Jamal ./for.out.jam -> ./for.out                                                                                                                                                                           
> javax0.jamal.api.BadSyntaxAt: The keyword 'in' is missing in the 'for' macro '$i in (bear                                                                                                                  
> zebra                                                                                                                                                                                                      
> snake                                                                                                                                                                                                      
> dog
> cat)=* $i
> ' at ./for.out.jam/1:24
>         at javax0.jamal.builtins.For.checkKeyword(For.java:137)
>         at javax0.jamal.builtins.For.evaluate(For.java:32)
>         at javax0.jamal.engine.Processor.evaluateBuiltinMacro(Processor.java:374)
>         at javax0.jamal.engine.Processor.lambda$evaluateBuiltInMacro$5(Processor.java:325)

Do I miss something ?
Regards.

**Reply by ypiel-talend** [reply]
Humm it seems an old version....
